### PR TITLE
Enhancements to DoublePrecision

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

File: numerics/double_precision_body.hpp [hunk 1]
```diff
@@ -19,14 +21,36 @@ constexpr DoublePrecision<T>::DoublePrecision(T const& value)
...
   value = temp + y;
   error = (temp - value) + y;
+  return *this;
+}
+
+template<typename T>
+DoublePrecision<T>& DoublePrecision<T>::operator+=(
+    DoublePrecision<Difference<T>> const& right) {
+  *this = *this + right;
+  return *this;
```
Comment (eggrobin) [inline numerics/double_precision_body.hpp]: why not `*this = *this + right`? Similarly below.

File: numerics/root_finders_body.hpp [hunk 2]
```diff
@@ -72,10 +72,9 @@ std::set<Argument> SolveQuadraticEquation(
 
   static Discriminant const discriminant_zero{};
 
-  // Use compensated summation for the discriminant because there can be
+  // Use double precision for the discriminant because there can be
   // cancellations.
-  DoublePrecision<Discriminant> discriminant(a1 * a1);
-  discriminant.Increment(-4.0 * a0 * a2);
+  DoublePrecision<Discriminant> discriminant = TwoSum(a1 * a1, -4.0 * a0 * a2);
```
Comment (eggrobin) [inline numerics/root_finders_body.hpp]: What does Higham actually do? if Higham only does compensated summation, mention the difference in the comment above.

File: numerics/double_precision.hpp [hunk 3]
```diff
@@ -10,16 +10,25 @@ namespace internal_double_precision {
...
+using quantities::Product;
+using quantities::Sum;
 
-// A simple container for accumulating a value using compensated summation.  The
-// type of the value must be an affine space.  The value constructor is not
-// explicit to make it easy to construct an object with no error.
+// A simple container for accumulating a value using double precision.  The
+// type of the value must be an affine space.  The notations follow
+// Library for Double-Double and Quad-Double Arithmetic, Hida, Li and Bailey,
+// 2007.
```
Comment (eggrobin) [inline numerics/double_precision.hpp]: I think the convention we mostly use is `authors (year), title`.

File: numerics/double_precision.hpp [hunk 4]
```diff
@@ -10,16 +10,25 @@ namespace internal_double_precision {
...
 template<typename T>
 struct DoublePrecision final {
   constexpr DoublePrecision() = default;
+
+  // This constructor is not explicit to make it easy to construct an object
+  // with no error.
   constexpr DoublePrecision(T const& value);  // NOLINT(runtime/explicit)
 
-  void Increment(Difference<T> const& increment);
+  DoublePrecision<T>& operator+=(Difference<T> const& right);
```
Comment (eggrobin) [inline numerics/double_precision.hpp]: Why is there no `operator+` corresponding to that?

File: numerics/root_finders_body.hpp (hunk 2 above)
Comment (pleroy) [inline numerics/root_finders_body.hpp]: Done.

File: numerics/double_precision.hpp (hunk 3 above)
Comment (pleroy) [inline numerics/double_precision.hpp]: Done.

File: numerics/double_precision_body.hpp (hunk 1 above)
Comment (pleroy) [inline numerics/double_precision_body.hpp]: Let's not go there, 'tis a silly place.

**Pull request merged**